### ISSUE TITLE
Starting from version 12 it is not possible to override the fetch API used by SDK

### ISSUE THREAD

**Issue by ivanBereznev** [realm/realm-js#6159]
### How frequently does the bug occur?

Always

### Description

Previously Node.js SDK was using `@realm/network-transport` package as a provider of fetch API and it was possible to override it and use e.g. `window.fetch` API when running in an Electron renderer process. But starting from version 12.0 the SDK seems to use `node-fetch` exclusively without an option to override it.
The option to use custom fetch API was important because it allowed to control SSL validation for network operations performed by SDK, for example usage of `window.fetch` in an Electron renderer allows to trust certificates added to the system certificate store.

I see there is an option to define a custom function for SSL validation in [Sync configuration](https://www.mongodb.com/docs/realm-sdks/js/latest/types/SSLConfiguration.html) which is a good feature that could add to flexibility, however there seems to be no way to control SSL validation for requests sent by SDK to log user in. This defeats the purpose of SSL settings in the sync configuration because the SSL validation issues it should help to resolve will prevent app from getting a user session needed to sync data down.

### Can you reproduce the bug?

Always

### Reproduction Steps

1. Run Realm in Electron, in a renderer process with node integration.
2. Try to inject `window.fetch` API into realm SDK in order to trust custom certificates from the system store.

### Version

12.1.0

### What services are you using?

Both Atlas Device Sync and Atlas App Services

### Are you using encryption?

No

### Platform OS and version(s)

Windows 10, Debian 11

**Comment by kneth** [comment]
@ivanBereznev Thank you for reporting. It is correct that the SSL configuration only affects Altas Device Sync connections, and logging in, calling Altas Functions, etc. are not. 

v12 is a rewrite of the SDK, and part of building, we are creating a bundle. The network transport is rolled up as we haven't considered the case where a custom `fetch` implementation was used. It is correct that for Electron, we are also using `node-fetch` as we don't have separate bundles for node and Electron.

We need to investigate how we temporarily (see sidenote below) can expose an API/configuration so you can pick your own network transport. To help to provide a useful API, may I ask you to share a code snippet on how you injected the transport prior to v12?

As a sidenote, we are working on relying on native/platform specific network implementations. I can't give you a timeline for when it will be production ready. The driving force is to make what you describe more accessible.

**Comment by ivanBereznev** [comment]
@kneth thank you for the quick response.

> To help to provide a useful API, may I ask you to share a code snippet on how you injected the transport prior to v12?

With SDK of version 10.24.0 that we are using now there is the following script executed as a part of build procedure:
```
const {readJsonSync, writeJsonSync} = require('fs-extra')

const realmNetPackage = readJsonSync('./node_modules/realm-network-transport/package.json')
realmNetPackage.main = './dist/bundle.dom.cjs.js'
writeJsonSync('./node_modules/realm-network-transport/package.json', realmNetPackage, {spaces: 2})
```
This is not the cleanest way of doing this, but it works. As far as I understand there is also an option to import the `DefaultNetworkTransport` from that package and assign `fetch` and `AbortController` to it before utilizing SDK functions that require network requests. Maybe in v12 SDK could export the `DefaultNetworkTransport` or provide some other way of injecting the fetch API.

> As a sidenote, we are working on relying on native/platform specific network implementations. I can't give you a timeline for when it will be production ready. The driving force is to make what you describe more accessible.

Ah, that's good news, sounds very promising.

**Comment by kneth** [comment]
@ivanBereznev In #6165 we have added a new configuration option which will make it possible to override `fetch`.

**Comment by ivanBereznev** [comment]
@kneth perfect, thank you! But what about `AbortController`? When I've tried to replace the `node-fetch` with some other package that exports `window.fetch` I ended up getting errors about an unrecognized signal which made me think that I have to inject the proper `AbortController` in addition to the `fetch`.

**Comment by kneth** [comment]
@ivanBereznev I have moved it back to draft state, and we can see how we can support custom abort controllers.

**Comment by kneth** [comment]
@ivanBereznev 

In #6165 we have exposed an option to use a custom `fetch` implementation. I wonder if you can add your custom AbortController to it:

```js
const HTTP_METHOD: Record<number, string> = {
  [0]: "GET",
  [1]: "POST",
  [2]: "PUT",
  [3]: "PATCH",
  [4]: "DELETE",
};

const customFetch = async (request: Request) => {
  const { url, body, ...rest } = request;
  if (typeof request.method == "number") {
    rest.method = HTTP_METHOD[request.method];
  }
  let abortController = new CustomAbortController();
  let signal = abortController.signal;
  const response = await fetch(url, { ...rest, signal });
  return response;
};
```

The custom `fetch` is applied as part of the app configuration:

```
let app = new Realm.App({ id: "footer", fetchOverride: customFetch });
```

**Comment by ivanBereznev** [comment]
@kneth sorry, I think we have a misunderstanding here so let me rephrase. 
I don't need to abort the requests myself, but in addition to the `DefaultNetworkTransport.fetch` I want to override the [DefaultNetworkTransport.AbortController](https://github.com/realm/realm-js/blob/main/packages/realm-network-transport/src/node/index.ts#L28) as well. Because I think it won't function properly if I'll inject the `window.fetch` only and then SDK will try to use that with the `AbortController` from node.

**Comment by takameyer** [comment]
@ivanBereznev sorry for the late reply.  I'll take a look at the proposed PR and try and find a way to override the `AbortController`

**Comment by takameyer** [comment]
@ivanBereznev In the PR we provided, it would now be possible to override `fetch` with a custom function (ie `window.fetch`).  Upon doing this, the `DefaultNetworkTransport` would become unused as well as the `AbortController` assigned to it.  Unless your custom `fetch` is calling `transport.fetch`, the solution @kneth provided should work.
Here is the [source](https://github.com/realm/realm-js/blob/main/packages/realm/src/platform/network.ts#L52-L60) of the original `fetch` which your could now completely override.

Does this meet your expectations?

**Comment by ivanBereznev** [comment]
@takameyer okay, if the abort controller becomes unused when the custom fetch is specified I believe the PR meets my expectations. Thanks a lot!

**Comment by kneth** [comment]
@ivanBereznev It should be possible from v12.6.0.
